docs/rfcs/0001-edge-computation-in-storage-side.md

```diff
@@ -0,0 +1,78 @@
+# Summary
+
+Proposal the `Edge Computation` component to complete the push-down computation. The main principle of it is to push down a part of plan to storage to execute.
+
+# Motivation
+
+1. Support more complex push-down computation
+2. Simplify the push-down interface
+3. Unify the behavior of graph computation and push-down computation by reuse the computation executor
+
+# Usage explanation
+
+Will expose the interface in storage RPC services.
+```
+enum ExecutorKind {
+     kLimit,
+     kFilter,
+     kAggregate,
+     ...
+}
+
+// Or we could reuse the origin Limit Plan Node
+// and implement Serialize/Deserialize interface of fbthrift for it
+struct LimitExecutor {
+     i64 offset,
+     i64 limit,
+}
+
+// More executor...
+
+struct Executor {
+     ExecutorKind kind,
+     Optional LimitExecutor,
+     Optional FilterExecutor,
+     Optional AggregateExecutor,
+     ...
+}
+
+struct EdgeComputer {
+     // The executor depends on previous executor
+     // And get output of previous
+     list<Executor>
+}
+
+struct GetNeighborsRequest {
+     // Origin fields ...
+     EdgeComputer edge_computer,
+}
+```
+
+# Design explanation
+
+In storage side:
+
+The `Edge Computation` will append to the origin plan in storage, e.g. Append push-down plan into `GetNeighbors` origin plan. So storage scan data firstly, then complete the push-down computation and response data finally.
```

Comment:
If the idea coms from coprocessor, as far as I know, both hbase and tidb is the query layer to generate the whole plan, not only part. The append is hard to control, for now there is only one output node for `go`, what if there are more than one (e.g. lookup may have more than one when we lookup where e1.c1 == 1 or e2.c1 == 1)?

IMO, either just use flag, or the whole plan should be generate in query engine.

Reply:
As I known the tikv coprocessor only support a subset of executor, so maybe it can't push down the whole plan to storage.

Reply:
Yes, I means a part of the whole plan generated by query layer.

Reply:
I understand your idea. This is quite interesting, imagine we generate the whole plan for storage in query layer not only part of them, many optimization will be easier (e.g. filter). Query layer only need to merge/reorder the executor in logic plan.

If we only want to push down a part of plan to storage, there is not a big difference with just do optimize in query layer as long as we support agg/sort/dedup... in each storage RPC interface. We could talk offline if necessary.

Reply:
IMHO, most TP scenarios do not need to push down the calculation(sql-like) on the graph side to the storage side, if the index is implemented well (IO will not become the most critical bottleneck). So the point is to enhance the ability of the index rather than just push down.

Reply:
Yes, but I found new issue in nearly development. As report in #2583 , the order of push-down executor is not determined in current implementation. Now we could resolve #2583 by do filter before limit, but if graph layer require semantic that limit first (or other case), we can't fit it now.

Reply:
The problem is that the capacity of current implementation is not complete.